File: scripts/build_dictionary.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""从 IELTS Word List.txt 解析并生成 dictionary.js"""
import re
import os

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(SCRIPT_DIR)
INPUT = os.path.join(ROOT, 'data/wordbooks/IELTS Word List.txt')
OUTPUT = os.path.join(ROOT, 'src/config/dictionary.js')

def escape(s):
    if s is None:
        return ''
    return str(s).replace('\\', '\\\\').replace('"', '\\"').replace('\n', ' ')

def parse_line(line):
    line = line.strip()
    if not line or line.startswith('Word List') or line.startswith('README') or line.startswith('《'):
        return None
    if line.startswith('以 ') or line.startswith('范洪滔') or re.match(r'^\d{4}-\d{2}-\d{2}$', line):
        return None
    word = ''
    phonetic = ''
    pos = 'n.'
    definition = ''
    # 找音标开始 / 或 [ 或 {
    m = re.search(r'\s+(/[^/]+/|\[[^\]]+\]|\{[^}]+\})\s+', line)
    if m:
        word = line[:m.start()].strip().rstrip('*')
        phonetic = m.group(1)
        rest = line[m.end():].strip()
        pos_m = re.match(r'^([nva\./]+\.?)\s+(.+)$', rest, re.I)
        if pos_m and len(pos_m.group(1)) <= 12:
            pos = pos_m.group(1).strip()
            definition = pos_m.group(2).strip()
        else:
            definition = rest
    else:
        parts = re.split(r'\s{2,}', line, maxsplit=1)
        if len(parts) >= 2:
            word = parts[0].strip().rstrip('*')
            rest = parts[1].strip()
            pos_m = re.match(r'^([nva\./]+\.?)\s+(.+)$', rest, re.I)
            if pos_m and len(pos_m.group(1)) <= 12:
                pos = pos_m.group(1).strip()
                definition = pos_m.group(2).strip()
            else:
                definition = rest
    if not word or len(word) < 2:
        return None
    key = word.lower().replace(' ', ' ')
    if not re.match(r"^[a-z\-'\s]+$", key, re.I) or len(key) > 40:
        return None
    return (key, [phonetic, pos or 'n.', definition or '（暂无释义）'])

def main():
    with open(INPUT, 'r', encoding='utf-8') as f:
        lines = f.readlines()
    dict_ = {}
    for line in lines:
        entry = parse_line(line)
        if entry:
            k, v = entry
            dict_[k] = v
    keys = sorted(dict_.keys())
    out = []
    out.append('/**')
    out.append(' * WordQuest 内置词典（含雅思词汇）')
    out.append(' * 词条数: ' + str(len(keys)))
    out.append(' * 格式：word → [phonetic, partOfSpeech, definition]')
    out.append(' */')
    out.append('')
    out.append('const DICTIONARY = {')
    for k in keys:
        v = dict_[k]
        out.append('  "' + escape(k) + '": ["' + escape(v[0]) + '", "' + escape(v[1]) + '", "' + escape(v[2]) + '"],')
    out.append('};')
    out.append('')
    out.append('function dictLookup(word) {')
    out.append('  if (!word) return null;')
    out.append('  const key = word.toLowerCase().trim();')
    out.append('  const entry = DICTIONARY[key];')
    out.append('  if (!entry) return null;')
    out.append('  return { word: key, phonetic: entry[0], partOfSpeech: entry[1], definition: entry[2] };')
    out.append('}')
    out.append('')
    out.append('function dictRandomWords(count) {')
    out.append('  const keys = Object.keys(DICTIONARY);')
    out.append('  const shuffled = keys.sort(() => Math.random() - 0.5);')
    out.append('  return shuffled.slice(0, count).map(w => ({')
    out.append('    word: w, phonetic: DICTIONARY[w][0], partOfSpeech: DICTIONARY[w][1], definition: DICTIONARY[w][2]')
    out.append('  }));')
    out.append('}')
    out.append('')
    out.append('if (typeof module !== \'undefined\' && module.exports) {')
    out.append('  module.exports = { DICTIONARY, dictLookup, dictRandomWords };')
    out.append('}')
    with open(OUTPUT, 'w', encoding='utf-8') as f:
        f.write('\n'.join(out))
    print('Wrote', len(keys), 'entries to', OUTPUT)

if __name__ == '__main__':
    main()

File: scripts/build-dictionary.js
/**
 * 从 data/wordbooks/IELTS Word List.txt 解析并生成 src/config/dictionary.js
 * 运行: node scripts/build-dictionary.js
 */
const fs = require('fs');
const path = require('path');

const inputPath = path.join(__dirname, '../data/wordbooks/IELTS Word List.txt');
const outputPath = path.join(__dirname, '../src/config/dictionary.js');

const raw = fs.readFileSync(inputPath, 'utf-8');
const lines = raw.split(/\r?\n/);
const dict = {};
let skipped = 0;

// 匹配音标：/.../ 或 [...] 或 {...}
const phoneticMatch = /^(\s*)(\/[^/]+\/|\[[^\]]+\]|\{[^}]+\})(\s+)(.+)$/;
// 词性缩写
const posMatch = /^(n\.|v\.|a\.|ad\.|adv\.|vt\.|vi\.|conj\.|prep\.|pron\.|excl\.|art\.|aux\.|int\.|pl\.)(\s+)(.*)$/i;

for (const line of lines) {
  const trimmed = line.trim();
  if (!trimmed || trimmed.startsWith('Word List') || trimmed.startsWith('README') || trimmed.startsWith('《') || trimmed.startsWith('以 ') || trimmed.startsWith('范洪滔') || /^\d{4}-\d{2}-\d{2}$/.test(trimmed)) {
    skipped++;
    continue;
  }
  const firstChar = trimmed.charAt(0);
  if (firstChar === '' || firstChar === '─' || trimmed === '根据对最后' || trimmed.startsWith('本人不对')) continue;

  let word = '';
  let phonetic = '';
  let partOfSpeech = 'n.';
  let definition = '';

  const phonStart = trimmed.search(/\s*(\/|\{|\[)/);
  if (phonStart > 0) {
    word = trimmed.slice(0, phonStart).trim().replace(/\*$/, '');
    const afterWord = trimmed.slice(phonStart).trim();
    const bracket = afterWord[0];
    const bracketEnd = bracket === '/' ? '/' : (bracket === '[' ? ']' : '}');
    const endIdx = afterWord.indexOf(bracketEnd, 1);
    if (endIdx > 0) {
      phonetic = afterWord.slice(0, endIdx + 1);
      let rest = afterWord.slice(endIdx + 1).trim();
      const posMatch = rest.match(/^([n\.v\.a\.ad\.vt\.vi\.conj\.prep\.pron\.excl\.aux\.int\.pl\.]+(?:\/[n\.v\.a\.]+\.?)*)\s+(.+)$/i) || rest.match(/^([nva]\.?)\s+(.+)$/i);
      if (posMatch && posMatch[1].length <= 15) {
        partOfSpeech = posMatch[1].trim();
        definition = posMatch[2].trim();
      } else {
        definition = rest;
      }
    } else {
      definition = afterWord;
    }
  } else {
    const twoSpaces = trimmed.indexOf('  ');
    if (twoSpaces > 0) {
      word = trimmed.slice(0, twoSpaces).trim().replace(/\*$/, '');
      let rest = trimmed.slice(twoSpaces).trim();
      const posMatch = rest.match(/^([n\.v\.a\.]+(?:\/[n\.v\.]+\.?)*)\s+(.+)$/i);
      if (posMatch && posMatch[1].length <= 10) {
        partOfSpeech = posMatch[1].trim();
        definition = posMatch[2].trim();
      } else {
        definition = rest;
      }
    }
  }

  if (!word || word.length < 2) continue;
  const key = word.toLowerCase().replace(/\s+/g, ' ');
  if (/^[a-z\-'\s]+$/i.test(key) && key.length <= 40) {
    dict[key] = [phonetic || '', partOfSpeech || 'n.', definition || '（暂无释义）'];
  }
}

// 读取现有 dictionary 并合并（IELTS 覆盖同键，其余保留）
let existing = {};
try {
  const existingContent = fs.readFileSync(outputPath, 'utf-8');
  const existingMatch = existingContent.match(/const DICTIONARY = \{([\s\S]*?)\};/);
  if (existingMatch) {
    const existingStr = '{' + existingMatch[1] + '}';
    existing = Function('return ' + existingStr)();
  }
} catch (_) {}
const merged = { ...existing, ...dict };

const keys = Object.keys(merged).sort();
const linesOut = [];
linesOut.push('/**');
linesOut.push(' * WordQuest 内置词典（含雅思词汇）');
linesOut.push(' * 词条数: ' + keys.length);
linesOut.push(' * 格式：word → [phonetic, partOfSpeech, definition]');
linesOut.push(' */');
linesOut.push('');
linesOut.push('const DICTIONARY = {');
function esc(s) {
  return (s || '').replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, ' ');
}
keys.forEach(k => {
  const v = merged[k];
  linesOut.push('  "' + esc(k) + '": ["' + esc(v[0]) + '", "' + esc(v[1]) + '", "' + esc(v[2]) + '"],');
});
linesOut.push('};');
linesOut.push('');
linesOut.push('function dictLookup(word) {');
linesOut.push('  if (!word) return null;');
linesOut.push('  const key = word.toLowerCase().trim();');
linesOut.push('  const entry = DICTIONARY[key];');
linesOut.push('  if (!entry) return null;');
linesOut.push('  return { word: key, phonetic: entry[0], partOfSpeech: entry[1], definition: entry[2] };');
linesOut.push('}');
linesOut.push('');
linesOut.push('function dictRandomWords(count) {');
linesOut.push('  const keys = Object.keys(DICTIONARY);');
linesOut.push('  const shuffled = keys.sort(() => Math.random() - 0.5);');
linesOut.push('  return shuffled.slice(0, count).map(w => ({');
linesOut.push('    word: w, phonetic: DICTIONARY[w][0], partOfSpeech: DICTIONARY[w][1], definition: DICTIONARY[w][2]');
linesOut.push('  }));');
linesOut.push('}');
linesOut.push('');
linesOut.push('if (typeof module !== \'undefined\' && module.exports) {');
linesOut.push('  module.exports = { DICTIONARY, dictLookup, dictRandomWords };');
linesOut.push('}');

fs.writeFileSync(outputPath, linesOut.join('\n'), 'utf-8');
console.log('Wrote', keys.length, 'entries to', outputPath);
